File: runtime/ruby/units/wasi/path_remove_directory.rb
# requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
def wasi_path_remove_directory(dirfd, path_ptr, path_len)
  rel = @memory.read_string(path_ptr, path_len)
  # rmdir(2) never follows a trailing symlink.
  host_path, err = resolve_path(dirfd, rel, follow_last: false)
  return err if err
  # rmdir through a trailing slash on an existing directory is EINVAL per
  # wasmtime; other shapes come from the host call.
  if host_path.end_with?("/") && File.directory?(host_path.delete_suffix("/"))
    return ERRNO_INVAL
  end
  Dir.rmdir(host_path)
  ERRNO_SUCCESS
rescue SystemCallError => e
  fs_errno(e)
end

File: runtime/perl/units/wasi/path_remove_directory.pl
# requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
sub wasi_path_remove_directory {
    my ($self, $dirfd, $path_ptr, $path_len) = @_;
    my $rel = $self->{memory}->read_string($path_ptr, $path_len);
    # rmdir(2) never follows a trailing symlink.
    my ($host, $err) = $self->resolve_path($dirfd, $rel, 0);
    return $err if defined $err;
    # rmdir through a trailing slash on an existing directory is EINVAL per
    # wasmtime (ADR-49); other shapes come from the host call.
    return ERRNO_INVAL if substr($host, -1) eq '/' && -d substr($host, 0, -1);
    rmdir($host) or return $self->fs_errno(0 + $!);
    return ERRNO_SUCCESS;
}
